File: RoomExample/app/src/main/java/com/example/sontbv/roomexample/MainActivity.java
package com.example.sontbv.roomexample;

import android.content.Intent;
import android.content.SharedPreferences;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Completable;
import io.reactivex.CompletableObserver;
import io.reactivex.Observable;
import io.reactivex.ObservableOnSubscribe;
import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Action;
import io.reactivex.observers.DisposableSingleObserver;
import io.reactivex.schedulers.Schedulers;

public class MainActivity extends AppCompatActivity {
    private final static String TAG = MainActivity.class.getSimpleName();
    CompositeDisposable compositeDisposable = new CompositeDisposable();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        final Repo repo = new Repo("rails123", "123", 1);

        final List<Repo> repos = new ArrayList<>();
        final User user = new User("sontbv", "sontbv");
        repos.add(new Repo("react", "xyz", 1));
        repos.add(new Repo("angular", "xyz", 1));
        repos.add(new Repo("vue", "xyz", 1));
        repos.add(new Repo("rails", "xyz", 1));
        repos.add(new Repo("php", "xyz", 1));

        if(isFirstTime()) {
            Completable.fromAction(new Action() {
                @Override
                public void run() throws Exception {
                    RepoDatabase.getInstance(getApplicationContext())
                            .getUserDao()
                            .insert(user);
                    RepoDatabase.getInstance(getApplicationContext())
                            .getRepoDao()
                            .insertListRecord(repos);
                }
            }).subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribeWith(new CompletableObserver() {
                        @Override
                        public void onSubscribe(Disposable d) {
                            Log.d(TAG, "subscribe");
                        }
                        @Override
                        public void onComplete() {
                            Log.d(TAG, TAG);
                        }
                        @Override
                        public void onError(Throwable e) {
                            Log.e(TAG, e.getMessage());
                        }
                    });
        } else {
            compositeDisposable.add(
                RepoDatabase.getInstance(getApplicationContext())
                    .getRepoDao()
                    .getAllRepos()
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribeWith(new DisposableSingleObserver<List<Repo>>() {
                        @Override
                        public void onSuccess(List<Repo> repos) {
                            Log.d(TAG, repos.size() + "");
                        }

                        @Override
                        public void onError(Throwable e) {
                            Log.e(TAG, e.getMessage());
                        }
                    })
            );

            compositeDisposable.add(
                RepoDatabase.getInstance(getApplicationContext())
                    .getRepoDao()
                    .findRepositoriesForUser(1)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribeWith(new DisposableSingleObserver<List<Repo>>() {
                        @Override
                        public void onSuccess(List<Repo> repos) {
                            Log.d(TAG, repos.size() + "");
                        }

                        @Override
                        public void onError(Throwable e) {
                            Log.e(TAG, e.getMessage());
                        }
                    })
            );


        }
    }

    private boolean isFirstTime(){
        Boolean isFirstTime = getSharedPreferences("PREFERENCE", MODE_PRIVATE)
                .getBoolean("isFirstTime", true);

        if(isFirstTime) {
            SharedPreferences.Editor editor = getSharedPreferences("PREFERENCE", MODE_PRIVATE).edit();
            editor.putBoolean("isFirstTime", false);
            editor.apply();
        }
        return isFirstTime;
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        compositeDisposable.clear();
    }

}

File: FlightTickets/app/src/main/java/com/example/sontbv/flighttickets/view/MainActivity.kt
package com.example.sontbv.flighttickets.view

import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.DefaultItemAnimator
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import com.example.sontbv.flighttickets.R
import com.example.sontbv.flighttickets.app.Const
import com.example.sontbv.flighttickets.network.ApiClient
import com.example.sontbv.flighttickets.network.ApiService
import com.example.sontbv.flighttickets.network.model.Price
import com.example.sontbv.flighttickets.network.model.Ticket
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.Observable;
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.functions.Function
import io.reactivex.schedulers.Schedulers

import kotlinx.android.synthetic.main.activity_main.*
import java.util.*
import io.reactivex.ObservableSource
import io.reactivex.internal.operators.single.SingleInternalHelper.toObservable
import io.reactivex.observables.ConnectableObservable
import io.reactivex.observers.DisposableObserver
import kotlin.collections.ArrayList


class MainActivity : AppCompatActivity() {
    private val TAG: String = MainActivity::class.java.simpleName
    private val from: String = "DEL"
    private val to: String = "HYD"

    private var disposable: CompositeDisposable = CompositeDisposable()
    private lateinit var apiService: ApiService
    private lateinit var adapter: TicketsAdapter
    private var tickets: MutableList<Ticket> = ArrayList()
    private lateinit var recyclerView: RecyclerView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        setSupportActionBar(toolbar)

        apiService = ApiClient.createService(ApiService::class.java)

        adapter = TicketsAdapter(this, tickets)
        recyclerView = findViewById(R.id.recycler_view)
        recyclerView.layoutManager = LinearLayoutManager(this)
        recyclerView.itemAnimator = DefaultItemAnimator()
        recyclerView.adapter = adapter

        var ticketsObservable: ConnectableObservable<MutableList<Ticket>>  = getTickets(from, to)!!.replay()

        disposable.add(
                ticketsObservable
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeWith(object : DisposableObserver<MutableList<Ticket>>() {
                            override fun onComplete() {

                            }

                            override fun onNext(t: MutableList<Ticket>) {
                                tickets.clear()
                                tickets.addAll(t)
                                adapter.notifyDataSetChanged()
                            }

                            override fun onError(e: Throwable) {
                            }
                        })
        )

        disposable.add(
                ticketsObservable
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .flatMap(object : Function<MutableList<Ticket>, ObservableSource<Ticket>> {
                            override fun apply(t: MutableList<Ticket>): ObservableSource<Ticket> {
                                return Observable.fromIterable(t)
                            }
                        })
                        .flatMap(object : Function<Ticket, ObservableSource<Ticket>> {
                            override fun apply(t: Ticket): ObservableSource<Ticket> {
                                return getPriceObservable(t)
                            }
                        })
                        .subscribeWith(object : DisposableObserver<Ticket>() {
                            override fun onComplete() {
                            }

                            override fun onNext(t: Ticket) {
                                var position = tickets.indexOf(t)
                                if(position == -1) {
                                    // TODO
                                    // Ticket not found in the list
                                    // This shouldn't happen
                                    return;
                                }
                            }

                            override fun onError(e: Throwable) {
                            }
                        })
        )




        ticketsObservable.connect()

    }


    private fun getTickets(from: String, to: String): Observable<MutableList<Ticket>>? {
        return apiService
                .searchTickets(from, to)
                .toObservable()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
    }
    // get price of flight
    private fun getPriceObservable(ticket: Ticket) : Observable<Ticket> {
        return apiService
                .getPrice(ticket.flightNumber, ticket.from, ticket.to)
                .toObservable()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .map(object: Function<Price, Ticket> {
                    override fun apply(t: Price): Ticket {
                        ticket.price = t
                        Log.d(TAG, t.price.toString())
                        return ticket
                    }
                })
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.action_settings -> true
            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        disposable.dispose()
    }
}
